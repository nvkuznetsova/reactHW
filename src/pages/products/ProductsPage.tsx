import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import React, { Component, ReactNode } from 'react';

import { badgeTexts } from 'src/constants/badgeTexts';
import { btnLabels } from 'src/constants/btnLabels';
import { Cart } from 'src/domain/Cart';
import { Product } from 'src/domain/Product';
import { CartActions } from 'src/store/cart/actions';
import { ProductsActions } from 'src/store/products/actions';
import { MainLayout } from 'src/components/MainLayout/MainLayout';
import { Button } from 'src/components/Button/Button';
import { 
  Card,
  CardBody,
  CardText,
  CardTitle,
} from 'src/components/Card';

const { available, unavailable } = badgeTexts;

export type ProductsProps = {
  products: ReadonlyArray<Product>,
  isLoading: boolean,
  error: Error | string,
  getProducts: () => ProductsActions,
  buyProduct: (cartItem: Cart) => CartActions,
};

export class ProductsPage extends Component<ProductsProps> {
  componentDidMount() {
    const { getProducts } = this.props;
    getProducts();
  }

  buyProductEvent(product: Product): void {
    const { buyProduct } = this.props;

    const cartItem: Cart = {
      ...product,
      count: 1,
    }
    buyProduct(cartItem);
  }

  renderCard(product: Product): ReactNode {
    const { isAvailable } = product;
    const cardClassName = isAvailable ? 'border-primary' : 'border-warning';

    return (
      <Card 
        cardClassName={cardClassName}
      > 
        { this.renderCardBody(product) }
      </ Card>
      )
  }

  renderCardBody(product: Product): ReactNode {
    const { name, isAvailable, price } = product;
    const badgeText = isAvailable ? available : unavailable;
    const badgeClassName = isAvailable ? 'badge-info' : 'badge-danger';
    const cardText = `Price: $${price}`;
    const btnType = isAvailable ? 'btn-primary' : 'btn-warning';
    const btnClassName = `w-100 ${btnType}`;

    return (
      <CardBody>
        <CardTitle 
          name={name.toLocaleUpperCase()}
          badgeText={badgeText}
          badgeClassName={badgeClassName} />
        <CardText cardText={cardText}/>
        <Button 
          btnLabel={btnLabels.buyButton}
          btnClassName={btnClassName}
          disabled={!isAvailable}
          click={this.buyProductEvent.bind(this, product)}
          />
      </CardBody>
    )
  }

  render() {
    const {
      products,
      isLoading,
      error,
    } = this.props;

    if (isLoading) {
      return <span>Loading</span>;
    }

    const hasError = !isLoading && !isEmpty(error);

    if (hasError) {
      const errorMessage = get(error, 'message', 'error');
      return <div>{errorMessage}</div>
    }

    return (
      <MainLayout>
        <div className="d-flex flex-wrap justify-content-evenly">
          {products.map((product, index) => {
            return (
              <div
                className="col-lg-4 col-md-6 align-self-center"
                key={index}
              >
                {this.renderCard(product)}
              </div>
            )
          })}
        </div>
      </MainLayout>)
  }
}