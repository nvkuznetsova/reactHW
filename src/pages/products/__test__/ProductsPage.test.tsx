import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { ProductsPage, ProductsProps } from '../ProductsPage';

describe('ProductsPage', () => {
  const product = {
    id: 1,
    name: "apple",
    description: "tasty",
    price: 1,
    category: "fruits",
    isAvailable: true,
  };
  const getProps = (): ProductsProps => ({
    products: [product],
    isLoading: false,
    error: '',
    getProducts: jest.fn(),
    buyProduct: jest.fn(),
  });

  const setup = (props = {}): ShallowWrapper => (
    shallow(<ProductsPage {...getProps()} {...props}/>)
  );

  it('should match snapshot', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('should match snapshot on loading', () => {
    const component = setup({ isLoading: true });
    expect(component).toMatchSnapshot();
  });

  it('should match snapshot on error', () => {
    const component = setup({ error: 'error' });
    expect(component).toMatchSnapshot();
    expect(component.find('div').text()).toBe('error');
  });

  it('should call `getProducts` on mount', () => {
    const getProducts =  jest.fn();
    setup({ getProducts });

    expect(getProducts).toHaveBeenCalled();
  });
});