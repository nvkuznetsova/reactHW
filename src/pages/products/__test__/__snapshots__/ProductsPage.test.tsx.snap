// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsPage should match snapshot 1`] = `
<MainLayout>
  <div
    className="d-flex flex-wrap justify-content-evenly"
  >
    <div
      className="col-lg-4 col-md-6 align-self-center"
      key="0"
    >
      <Card
        cardClassName="border-primary"
      >
        <CardBody>
          <CardTitle
            badgeClassName="badge-info"
            badgeText="Available"
            name="APPLE"
          />
          <CardText
            cardText="Price: $1"
          />
          <Button
            btnClassName="w-100 btn-primary"
            btnLabel="Buy"
            click={[Function]}
            disabled={false}
          />
        </CardBody>
      </Card>
    </div>
  </div>
</MainLayout>
`;

exports[`ProductsPage should match snapshot on error 1`] = `
<div>
  error
</div>
`;

exports[`ProductsPage should match snapshot on loading 1`] = `
<span>
  Loading
</span>
`;
